sql/logic_test.go

```diff
@@ -0,0 +1,321 @@
+
+		case "query":
+			query := logicQuery{pos: fmt.Sprintf("%s:%d", base, s.line)}
+			// Parse "query error <regexp>"
+			if m := errorRE.FindStringSubmatch(s.Text()); len(m) == 2 {
+				query.expectErr = m[1]
+			} else {
+				// TODO(pmattis): Parse "query <type-string> <sort-mode> <label>". The
+				// type string specifies the number of columns and their types: T for
+				// text, I for integer and R for floating point. The sort mode is one
+				// of "nosort", "rowsort" or "valuesort". The default is "nosort".
+				//
+				// The label is optional. If specified, the test runner stores a hash
+				// of the results of the query under the given label. If the label is
+				// reused, the test runner verifieds that the results are the
+				// same. This can be used to verify that two or more queries in the
+				// same test script that are logically equivalent always generate the
+				// same output.
+			}
+			var buf bytes.Buffer
+			for s.Scan() {
+				t := s.Text()
+				if t == "----" {
+					break
+				}
+				fmt.Fprintln(&buf, s.Text())
+			}
+			query.sql = strings.TrimSpace(buf.String())
+
+			// Query results are either a space separated list of values up to a
+			// blank line or a line of the form "xx values hashing to yyy". The
+			// latter format is used by sqllogictest when a large number of results
+			// match the query.
+			if !s.Scan() {
+				t.Fatalf("%d: expected query results line\n", s.line)
+			}
+			if m := resultsRE.FindStringSubmatch(s.Text()); len(m) == 3 {
+				var err error
+				query.expectedValues, err = strconv.Atoi(m[1])
+				if err != nil {
+					t.Fatal(err)
+				}
+				query.expectedHash = m[2]
+			} else {
+				for {
+					results := strings.Fields(s.Text())
+					if len(results) == 0 {
+						break
+					}
+					query.expectedResults = append(query.expectedResults, results...)
+					if !s.Scan() {
+						break
+					}
+				}
+				query.expectedValues = len(query.expectedResults)
+			}
+
+			t.execQuery(db, query)
+
+		case "halt", "skipif", "onlyif":
+			t.Fatalf("unimplemented test statement: %s", s.Text())
+		}
+	}
+
+	if err := s.Err(); err != nil {
+		t.Fatal(err)
+	}
+}
+
+func (t logicTest) execStatement(db *sql.DB, stmt logicStatement) {
+	fmt.Printf("%s: %s\n", stmt.pos, stmt.sql)
+	_, err := db.Exec(stmt.sql)
+	switch {
+	case stmt.expectErr == "":
+		if err != nil {
+			t.Fatalf("%s: expected success, but found %v", stmt.pos, err)
+		}
+	case !testutils.IsError(err, stmt.expectErr):
+		t.Fatalf("%s: expected %s, but found %v", stmt.pos, stmt.expectErr, err)
+	}
+}
+
+func (t logicTest) execQuery(db *sql.DB, query logicQuery) {
+	fmt.Printf("%s: %s\n", query.pos, query.sql)
+	rows, err := db.Query(query.sql)
+	switch {
+	case query.expectErr == "":
+		if err != nil {
+			t.Fatalf("%s: expected success, but found %v", query.pos, err)
+		}
+	case !testutils.IsError(err, query.expectErr):
+		t.Fatalf("%s: expected %s, but found %v", query.pos, query.expectErr, err)
+	default:
+		return
```

Comment:
It's mildly confusing. If there wasn't an error we would have fallen out of the switch above in the `case.expectErr == ""`.